### PR TITLE
RavenDB-18938 Track stack size inside LINQ and warn the user against potential StackoverflowException

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18938 

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 1]
```diff
@@ -220,6 +223,21 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
...
         }
 
+        internal void CheckDepthOfStackInOutputMap(IndexDefinition indexMetadata, DocumentDatabase documentDatabase)
+        {
+            var performanceHintConfig = documentDatabase.Configuration.PerformanceHints;
+            if (StackSizeInSelectClause > performanceHintConfig.MaxDepthOfRecursionInLinqSelect)
+            {
+                documentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                    documentDatabase.Name,
+                    $"Index '{indexMetadata.Name}' contains a lot of `let` clause.",
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]:
```suggestion
                    $"Index '{indexMetadata.Name}' contains a lot of `let` clauses.",
```

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 2]
```diff
@@ -220,6 +223,21 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
...
+                    $"Each of let clause is nesting projections and could potentially leads to StackoverflowException.",
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]:
```suggestion
                    $"Each of the let clause is nesting projections and could potentially leads to StackoverflowException.",
```

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: Need to also include a proper fix for that. Include a sample of a bad index and a transformation for a good one

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 3]
```diff
@@ -220,6 +223,21 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
             funcs.Add(map);
         }
 
+        internal void CheckDepthOfStackInOutputMap(IndexDefinition indexMetadata, DocumentDatabase documentDatabase)
+        {
+            var performanceHintConfig = documentDatabase.Configuration.PerformanceHints;
+            if (StackSizeInSelectClause > performanceHintConfig.MaxDepthOfRecursionInLinqSelect)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:233]: I think we want two settings:
* > 20 - we raise an alert
* > 50 - we refuse to allow the index (with the appropriate notice on how to configure that, of course)

**Comment by grisha-kotler** [comment]
Can it be moved to 5.2?

File: src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs [hunk 4]
```diff
@@ -850,11 +858,12 @@ private static StatementSyntax HandleReduce(string reduce, FieldNamesValidator f
             var rewrittenExpression = (CSharpSyntaxNode)mapRewriter.Visit(expression);
 
             StatementSyntax optimized = null;
-
+            bool? foundSourceDocumentInProjection;
             try
             {
                 var visitor = new RavenLinqOptimizer(fieldValidator);
                 optimized = visitor.Visit(new RavenLinqPrettifier().Visit(rewrittenExpression)) as StatementSyntax;
+                foundSourceDocumentInProjection = visitor.FoundSourceDocumentInProjection;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs]: It looks like leftover from another work and it looks that it makes compilation errors

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:233]: Refusing to put an index is something that might be a breaking change for someone. This worries me. Consider the case that a new node is added to the cluster - then an index with such problem could not be put.

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: Including a sample can be a bit problematic due to lack of space and proper formatting for the notifications in the studio. I think we already have a section in the docs where we explain the indexing alerts.

File: test/SlowTests/Issues/RavenDB_18938.cs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+using Raven.Client.Documents.Indexes;
+using Raven.Server.NotificationCenter.Notifications;
+using Sparrow.Json.Parsing;
+using Sparrow.Server.Collections;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Issues;
+
+public class RavenDB_18938 : RavenTestBase
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_18938.cs:15]: Would it alert on the index from https://issues.hibernatingrhinos.com/issue/RavenDB-15476 as well?

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 6]
```diff
@@ -220,6 +223,21 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
...
+            {
+                documentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                    documentDatabase.Name,
+                    $"Index '{indexMetadata.Name}' contains a lot of `let` clause.",
+                    $"Each of let clause is nesting projections and could potentially leads to StackoverflowException.",
+                    PerformanceHintType.Indexing,
+                    NotificationSeverity.Info,
+                    nameof(IndexCompiler)));
+            }
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:245]: Can we also put that warning into the log?

File: test/SlowTests/Issues/RavenDB_18938.cs (hunk 5 above)
Comment (maciejaszyk) [reply test/SlowTests/Issues/RavenDB_18938.cs:15]: No. It has a depth about `this21.[..].this0` but still SOE happens.

File: src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs]: done

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: done

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:245]: done

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 7]
```diff
@@ -220,6 +227,23 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
...
+            {
+                documentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                    documentDatabase.Name,
+                    $"Index '{indexMetadata.Name}' contains a lot of `let` clauses.",
+                    $"Each of the let clause is nesting projections and could potentially leads to StackoverflowException.",
+                    PerformanceHintType.Indexing,
+                    NotificationSeverity.Info,
+                    nameof(IndexCompiler)));
+                
+                Log.Info($"Index '{indexMetadata.Name}' contains a lot of `let` clauses.");
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: 1. Let's change it to Operations level (it's turned on by default). Info level is much more detailed and it's supposed to be enabled on demand if needed
2. Before doing the actual logging we always check if the logger itself is gonna do it so please add `if` as below:

```
if (Log.IsOperationsEnabled)
    Log.Operations(...)
```

3. I wonder if we shouldn't include `StackSizeInSelectClause` number here as well

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs [hunk 8]
```diff
@@ -507,19 +507,24 @@ static void RegisterPackage(NuGetFetcher.NuGetPackage package, bool userDefined,
...
-
+            var maxDepthInRecursiveLinqQuery = 0;
+            
             for (var i = 0; i < maps.Count; i++)
             {
                 var map = maps[i];
-                statements.AddRange(HandleMap(definition.SourceType, map, fieldNamesValidator, methodDetector, ref members));
+                statements.AddRange(HandleMap(definition.SourceType, map, fieldNamesValidator, methodDetector, stackDepthRetriever, ref members));
+
+                maxDepthInRecursiveLinqQuery = Math.Max(maxDepthInRecursiveLinqQuery, stackDepthRetriever.StackSize);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs:526]: shouldn't you also account here for the `reduce` as well?

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 9]
```diff
@@ -220,6 +227,24 @@ protected void AddMapInternal(string collection, string subCollecction, Indexing
...
         }
 
+        internal void CheckDepthOfStackInOutputMap(IndexDefinition indexMetadata, DocumentDatabase documentDatabase)
+        {
+            var performanceHintConfig = documentDatabase.Configuration.PerformanceHints;
+            if (StackSizeInSelectClause > performanceHintConfig.MaxDepthOfRecursionInLinqSelect)
+            {
+                documentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                    documentDatabase.Name,
+                    $"Index '{indexMetadata.Name}' contains a lot of `let` clauses.",
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: Add the actual number here.

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 6 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:245]: Add the number here.
And the amount you exceeded it by.

Do we have a way to add a link to docs?

File: src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs (hunk 8 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs:526]: Done

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:245]: Included. 
About the docs: maybe in `msg` part?

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 9 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: done

**Comment by ppekrol** [comment]
@maciejaszyk Can you rebase?

**Comment by maciejaszyk** [comment]
> @maciejaszyk Can you rebase?

To `5.3` or `5.2`?

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:245]: First we need to write about it in the docs. Then we can use the docs link generator and add it to the message.

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**